docs/design.md

```diff
@@ -424,29 +424,39 @@ supplied to provide an upper bound on timestamps for already-committed
 data (`ε` is the maximum clock skew). As the transaction progresses, any
 data read which have timestamps greater than `t` but less than `t+ε`
 cause the transaction to abort and retry with the conflicting timestamp
-t<sub>c</sub>, where t<sub>c</sub> \> t. The maximum timestamp `t+ε` remains the same.
-Time spent retrying because of reading recently committed data has an
-upper bound of `ε`. In fact, this is further optimized: upon restarting,
-the transaction not only takes into account the timestamp of the future
-value, but the timestamp of the node at the time of the uncertain read.
-The larger of those two timestamps (typically, the latter) is used to
-bump up the read timestamp, and additionally the node is marked as
-“certain”. This means that for future reads to that node within the
-transaction, we can set `MaxTimestamp = Read Timestamp` (and hence avoid
-further uncertainty restarts). Correctness follows from the fact that we
+t<sub>c</sub>, where t<sub>c</sub> \> t. The maximum timestamp `t+ε` remains
+the same.
+
+We apply another optimization to avoid further restarts caused
+by uncertainty. Upon restarting, the transaction not only takes
+into account t<sub>c<sub>, but the timestamp of the node at the time
+of the uncertain read t<>node<>. The larger of those two timestamps:
+max(t<sub>c<sub>, t<sub>node<sub>) (more likely the latter) is used
+to bump up the read timestamp. The use of t<sub>node<sub> guarantees that the
+time spent retrying because of reading recently committed data has an
+upper bound of `ε`.
+
+Additionally, the conflicting node is marked as “certain”. This means that
+for future reads to that node within the transaction, we can set
+`MaxTimestamp = Read Timestamp`. Correctness follows from the fact that we
 know that at the time of the read, there exists no version of any key on
-that node with a higher timestamp; if we ran into one during a future
-read, that node would have happened (in absolute time) after our
-transaction started.
+that node with a higher timestamp that was updated in the past in absolute
+time. A higher timestamp is indicative that the update happened in the future
+in absolute time, and the transaction will move forward reading an older
+version of the data (at the transactions timestamp). (This seems wrong; it only
+guarantees that a transaction coordinated from this node cannot have a greater
+timestamp. You can still have transactions coordinated from other nodes with
```

Comment:
> But what's missing in the discussion is the existence of a reader cache on each node that keeps track of recently read local values.

We call it the `(read) timestamp cache`. I'm not sure how much it pertains to the above situation because while it will move the txn timestamp forward, it'll still trail the node's time. But it's mentioned only briefly in [Lock-free distributed txns](https://github.com/cockroachdb/cockroach/blob/master/docs/design.md#lock-free-distributed-transactions) and certainly deserves more airplay, especially if its implications seem unclear.